File: app/src/main/java/com/sakurafish/pockettoushituryou/view/fragment/AdBannerFragment.java
package com.sakurafish.pockettoushituryou.view.fragment;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.os.Bundle;
import android.os.Handler;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.localbroadcastmanager.content.LocalBroadcastManager;

import com.google.android.gms.ads.AdListener;
import com.sakurafish.pockettoushituryou.databinding.FragmentAdbannerBinding;
import com.sakurafish.pockettoushituryou.di.Injectable;
import com.sakurafish.pockettoushituryou.view.helper.AdsHelper;

import javax.inject.Inject;

import static com.sakurafish.pockettoushituryou.view.helper.AdsHelper.ACTION_BANNER_CLICK;
import static com.sakurafish.pockettoushituryou.view.helper.AdsHelper.INTENT_EXTRAS_KEY_CLASS;

public class AdBannerFragment extends Fragment implements Injectable {
    private static final String TAG = AdBannerFragment.class.getSimpleName();

    private FragmentAdbannerBinding binding;
    private final Handler handler = new Handler();
    private Runnable runnable;
    private String parentActivityName;

    private BroadcastReceiver receiverADClick = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            if (intent.getAction() == null
                    || !intent.getAction().equals(ACTION_BANNER_CLICK)
                    || intent.getStringExtra(INTENT_EXTRAS_KEY_CLASS).equals(parentActivityName)) {
                return;
            }
            finishInterval();
            startInterval(false);
        }
    };

    @Inject
    AdsHelper adsHelper;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        binding = FragmentAdbannerBinding.inflate(inflater, container, false);
        return binding.getRoot();
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

        parentActivityName = getActivity().getClass().getSimpleName();
        setupAdView();
    }

    public void setupAdView() {
        if (adsHelper.isIntervalOK()) {
            binding.adView.setVisibility(View.VISIBLE);
            binding.adView.loadAd(adsHelper.getAdRequest());
        } else {
            startInterval(false);
        }

        binding.adView.setAdListener(new AdListener() {
            @Override
            public void onAdOpened() {
                startInterval(true);
            }
        });

        setADClickReceiver();
    }

    private void setADClickReceiver() {
        IntentFilter filter = new IntentFilter();
        filter.addAction(ACTION_BANNER_CLICK);
        LocalBroadcastManager.getInstance(getActivity()).registerReceiver(receiverADClick, filter);
    }

    private void startInterval(final boolean clicked) {
        binding.adView.setVisibility(View.GONE);

        long now = System.currentTimeMillis();

        if (clicked) {
            adsHelper.setLastClickTimeMillis(now);
            Intent localIntent = new Intent(ACTION_BANNER_CLICK);
            localIntent.putExtra(INTENT_EXTRAS_KEY_CLASS, parentActivityName);
            LocalBroadcastManager.getInstance(getActivity()).sendBroadcast(localIntent);
        }
        long intervalTimeMillis = adsHelper.getIntervalTimeMillis(now, adsHelper.getLastClickTimeMillis());

        runnable = () -> {
            binding.adView.setVisibility(View.VISIBLE);
            binding.adView.loadAd(adsHelper.getAdRequest());
        };
        handler.postDelayed(runnable, intervalTimeMillis);
    }

    private void finishInterval() {
        handler.removeCallbacks(runnable);
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        finishInterval();
        LocalBroadcastManager.getInstance(getActivity()).unregisterReceiver(receiverADClick);
    }
}

File: app/src/main/java/com/sakurafish/pockettoushituryou/view/fragment/AdBannerFragment.kt
package com.sakurafish.pockettoushituryou.view.fragment

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.annotation.WorkerThread
import androidx.fragment.app.Fragment
import androidx.localbroadcastmanager.content.LocalBroadcastManager
import com.google.android.gms.ads.AdListener
import com.sakurafish.pockettoushituryou.databinding.FragmentAdbannerBinding
import com.sakurafish.pockettoushituryou.di.Injectable
import com.sakurafish.pockettoushituryou.view.helper.AdsHelper
import com.sakurafish.pockettoushituryou.view.helper.AdsHelper.Companion.ACTION_BANNER_CLICK
import com.sakurafish.pockettoushituryou.view.helper.AdsHelper.Companion.INTENT_EXTRAS_KEY_CLASS
import kotlinx.coroutines.*
import javax.inject.Inject

class AdBannerFragment : Fragment(), Injectable {

    @Inject
    lateinit var adsHelper: AdsHelper

    private var _binding: FragmentAdbannerBinding? = null
    private val binding get() = _binding!!

    private lateinit var job: Job
    private var parentActivityName: String? = null

    private val receiverADClick = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            if (intent.action == null
                    || intent.action != ACTION_BANNER_CLICK
                    || intent.getStringExtra(INTENT_EXTRAS_KEY_CLASS) == parentActivityName) {
                return
            }
            finishInterval()
            startInterval(false)
        }
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        _binding = FragmentAdbannerBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)

        parentActivityName = requireActivity().javaClass.simpleName
        setupAdView()
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    private fun setupAdView() {
        if (adsHelper.isIntervalOK) {
            binding.adView.visibility = View.VISIBLE
            binding.adView.loadAd(adsHelper.adRequest)
        } else {
            startInterval(false)
        }

        binding.adView.adListener = object : AdListener() {
            override fun onAdOpened() {
                startInterval(true)
            }
        }
        setADClickReceiver()
    }

    private fun setADClickReceiver() {
        val filter = IntentFilter()
        filter.addAction(ACTION_BANNER_CLICK)
        LocalBroadcastManager.getInstance(requireContext()).registerReceiver(receiverADClick, filter)
    }

    private fun startInterval(clicked: Boolean) {
        binding.adView.visibility = View.GONE

        val now = System.currentTimeMillis()

        if (clicked) {
            adsHelper.lastClickTimeMillis = now
            val localIntent = Intent(ACTION_BANNER_CLICK)
            localIntent.putExtra(INTENT_EXTRAS_KEY_CLASS, parentActivityName)
            LocalBroadcastManager.getInstance(requireContext()).sendBroadcast(localIntent)
        }
        val intervalTimeMillis = adsHelper.getIntervalTimeMillis(now, adsHelper.lastClickTimeMillis)

        job = GlobalScope.launch {
            waitInterval(intervalTimeMillis)
        }
    }

    @WorkerThread
    suspend fun waitInterval(intervalTimeMillis: Long) {
        withContext(Dispatchers.IO) {
            delay(intervalTimeMillis)
            withContext(Dispatchers.Main) {
                binding.adView.visibility = View.VISIBLE
                binding.adView.loadAd(adsHelper.adRequest)
            }
        }
    }

    private fun finishInterval() {
        if (::job.isInitialized && job.isActive) job.cancel()
    }

    override fun onDestroy() {
        super.onDestroy()
        finishInterval()
        LocalBroadcastManager.getInstance(requireContext()).unregisterReceiver(receiverADClick)
    }
}
